### PR TITLE
[10.0] [ADD] hr_payroll_cancel

### PR DESCRIPTION
**Pull request by nikul-serpentcs**
HR Payroll Cancel
=================

This module allows the user to cancel a payslip whatever the previous state is without doing a refund. When the user cancels the journal entry is deleted and  the payslip state is set to reject. Then the user is able to set the state to draft again and later on he/she is able to confirm again the payslip.

If there’s a refund for a payslip the user should not cancel the entry because the refund would still be confirmed. In that case, the user has either to confirm again the payslip or cancel the refund.

### CODE REVIEW

**Comment by nikul-serpentcs** [comment]
@aheficent Could you please review.

**Review by AaronHForgeFlow — CHANGES_REQUESTED** [review]
Cannot cancel a payslip when is confirmed:

image

**Review by AaronHForgeFlow — CHANGES_REQUESTED** [review]
You should cancel the accounting entry created:
image

Besides, when there's a refund the user should not cancel the payslip. That was the behavior in v9.

**Review by AaronHForgeFlow — CHANGES_REQUESTED** [review]
I still can cancel a payslip when there's a refund. I should not be able.

**Comment by nikul-serpentcs** [comment]
@aheficent Could you check raise error if refund payslip is cancel.

File: hr_payroll_cancel/models/hr_payroll.py
Comment (AaronHForgeFlow) [inline hr_payroll_cancel/models/hr_payroll.py]: The credit_note flag is activated in the refund not in the original payslip. I am still able to cancel a payslip that can be refunded.

**Review by AaronHForgeFlow — CHANGES_REQUESTED** [review]

**Comment by nikul-serpentcs** [comment]
@aheficent Could you please review.

**Comment by nikul-serpentcs** [comment]
@aheficent Could you please review.

File: hr_payroll_cancel/__manifest__.py
Comment (feketemihai) [inline hr_payroll_cancel/__manifest__.py]: I think it should depend on hr_payroll_account, since it depends only on hr_payroll no account_move is created, so you don't have to cancel anything...

**Comment by feketemihai** [comment]
@nikul-serpentcs The module is ok, only i have a remark, if i hit by mistake the Refund button, i cannot cancel none of the payslips, so as a procedure i can cancel first the refund and after the initial payslip, because i don't want to see the balances in the accounts...

**Comment by AaronHForgeFlow** [comment]
@feketemihai I agree, the user may want to cancel the refund and then the payslip. But the user should not be able to cancel a payslip before canceling the refund. @nikul-serpentcs can you do the thing?

**Comment by aheficent** [comment]
@feketemihai I agree with you. The user should be able to cancel the refund first and then the payslip but not the other way around. @nikul-serpentcs can you do the thing?

**Comment by nikul-serpentcs** [comment]
@aheficent @feketemihai Improved code Could you please review.

**Review by feketemihai — CHANGES_REQUESTED** [review]

**Comment by nikul-serpentcs** [comment]
@aheficent @feketemihai Improved dependence Could you please review.

**Review by AaronHForgeFlow — COMMENTED** [review]
It seems that the related journal entry is not cancelled as well, can you check?

**Comment by nikul-serpentcs** [comment]
@aheficent Could you please review.

File: hr_payroll_cancel/models/hr_payroll.py
Comment (feketemihai) [inline hr_payroll_cancel/models/hr_payroll.py:34]: @nikul-serpentcs Can you please unlink the move, when it's cancelled all the move lines are deleted but the move remains as unposted, when you revalidate the payslip a new move is creeated, so i tend to unlink the move after it is canceled. Otherwise, work ok .

**Review by feketemihai — CHANGES_REQUESTED** [review]

**Comment by nikul-serpentcs** [comment]
@feketemihai @aheficent The latest code unlink the moves from the canceled payslip.
Could you please review it ??

**Comment by nikul-serpentcs** [comment]
Fixes leave overlap issues. #425

**Review by feketemihai — APPROVED** [review]

**Review by AaronHForgeFlow — COMMENTED** [review]
There's still an issue here:
I created a payslip and a refund. Then I cancelled the refund. Then tried to cancel the original payslip but got the following:

image

Showing the warning is fine IMO, but the message is misleading because the refund is already cancelled. IMO it should be, "please delete first the refund" or something similar.

**Review by AaronHForgeFlow — APPROVED** [review]
Please dismiss my last review. I was setting the refund to draft, not just cancelling it.  In that case it's correct to show a warning. Thus, :+1:

**Comment by AaronHForgeFlow** [comment]
Please fix the commit history before this is merged. 73 commits looks incorrect

**Comment by nikul-serpentcs** [comment]
@aheficent Fixed commit history, Can you Please check it ??

**Comment by AaronHForgeFlow** [comment]
@nikul-serpentcs IMO history is not fixed. This module is merged in v9 and all the history of that branch is missed here.

**Comment by nikul-serpentcs** [comment]
@aheficent Fixed commit history, Can you Please check it ??

**Comment by AaronHForgeFlow** [comment]
@nikul-serpentcs You're getting closer, the problem now it that your changes are missing. If you see commit 57b3c479fb93520bca6e875c7316163969646db0 your changes for the migration to v10 are missing

**Comment by AaronHForgeFlow** [comment]
I think you got it!  :medal_sports:

**Pull request merged**
